text/0015-status-update.md

```diff
@@ -0,0 +1,397 @@
+
+### a design for mutability
+
+My main concern right now is figuring out a model for compilation & mutability.
+For compilation, I'm currently imagining generating builder types, but I haven't
+dug into this yet.
+
+Offset resolution and rewriting will also be an important consideration. A
+compiler author should be able to use these tools to do things like deduplicate
+identical subtables, or to reorder tables as needed, which means we do need a
+mechanism for rewriting offsets as needed. There are various options here; for
+instance we could track offsets by ids, and then keep track of the location of
+the offsets as we write tables into a shared buffer, rewriting them later as
+needed. This will need more thought, experimentation, and discussion.
+
+### evaluating current performance
+
+It may also be useful to get some concrete numbers on the efficiency of the
+current work-to-date. I have spent a bit of time looking at disassembly and am
+generally confident that we should be able to achieve good results, but I
+haven't *demonstrated* this.
+
+I'm not sure this is strictly necessary at this point, but it might be helpful;
+if performance is actually terrible for some unforeseen reason, we might decide
+that our approach needs to be fundamentally rethought.
+
+I've also resisted this because I'm not sure of exactly what we should measure.
+In any case, it is something we will want to do at some point.
+
+### further refinements
+
+Finally, there are various additional features and functionality that it might
+be worth exploring, such as improving type information related to offsets (for
+instance encoding in the type of the offset what sort of table it should
+produce, or what table it is indexed into.) In addition, I am *very curious*
+about looking into migrating from doing codegen in a proc-macro to doing codegen
+as a separate build script that outputs normal rust files that are then included
+and used like any other. This has several advantages: it should reduce
+compilation time, but it should also produce more readable code: currently we
+generate lots of methods and type definitions that do not actually exist in any
+of our source files. This makes it harder for users to understand what is going
+on, and it also confuses tools like IDEs, which often cannot provide
+autocomplete for these generated methods.
+
+## To discuss
+
+- Is it worth doing some real-world profiling now? For instance, we could pretty
+  easily implement something like "iterate all glyphs and manually compute
+  bounding boxes", and we could try and compare that against HarfBuzz as well as
+  other Rust font parsing implementations.
+- Do my thoughts about hb-subset make sense? Put another way: would we be happy
+  with a result where subsetting is, say 50% slower than hb-subset (and I think
+  I'm being very pessimistic with that number), but is easier to maintain and
+  contribute to? This would still be orders of magnitude faster than, say,
+  fonttools.
```

Comment:
This imagined slowness would only exist in the subsetting case (basically we might need to do more intermediate allocation of subtables, instead of using a single big buffer) and wouldn't apply to shaping.

Reply:
QQ, can we get a mutable struct from a mutable u8's buffer w/o allocation?

Reply:
Without digging too deep into this, my instinct here is that the process of generating new subtables based on the closure can be batched and potentially parallelized. Since repacking handles ordering, we don't need to care about where we emit the bytes for these tables as we're producing them. I can see a fork/join model where we dispatch sets of tables to multiple threads for the actual subsetting and each thread dumps the results linearly into a local buffer. We accumulate the list of these buffers at the join and reference them following reordering to generate the final serialized buffer. This is also the point where we would patch the offsets. Note that any unmodified tables can simply reference the original font data.

This seems optimally time and space efficient, and shouldn't require aliased mutable borrows.

Reply:
In comparing notes with Behdad it seemed likely the algorithms are what matter, a few copies won't hurt us. We'll have to prove it ofc but I'm pretty optimistic.